course3/course_en.md

```diff
@@ -0,0 +1,724 @@
+```
+
+In `Option[T]`, the notation `[T]` indicates that `Option` is a generic type, and the value it holds is of type `T`. For example:
+- `Option[Int]`: it can either hold a value of type `Int` or it can be empty.
+
+We will explore how to extract the value from `Some` shortly.
+
+---
+
+# Definition of Local Functions
+
+<style scoped>
+    .two-columns { columns: 2; }
+</style>
+
+<div class="two-columns">
+In most cases, the type and return type of a local function can be omitted, and even its name can also be omitted.
+
+In MoonBit, functions are considered "first-class citizens", i.e., they can be used as parameters, return values, and can also be assigned or stored.
+
+```moonbit expr
+let answer: () -> Int = fn () {
+  fn real_answer(i) {
+    42
+  }
+  real_answer("Ultimate Question")
+}
+
+let x: Int = answer() // 42
+```
+
+![height:70px](../pics/func_without_parameter_type.png)
+</div>
+
+---
+
+# Function Types
+
+`(<param type>, <param type>, <param type>, ...) -> <return type>`
+
+- `() -> Int`
+- `(Int, String, Char) -> Int`
+- `((Int, Int, Int)) -> (Int, Int, Int)` accepts a tuple and returns a tuple
+
+---
+
+# Data Structure: Lists
+
+---
+
+# List: A Sequence of Data
+
+- Sometimes, we have data with the following characteristics:
+    - The data is ordered.
+    - The data can be duplicated.
+    - The quantity of data is uncertain.
+- For example:
+    - Words in a sentence: [ `"Words"` `"in"` `"a"` `"sentence"` ]
+    - DNA sequence: [`G` `A` `T` `T` `A` `C` `A`]
+    - ...
+
+---
+
+# List Interfaces
+
+Taking a single-ended immutable integer list `IntList` as an example, the following operations should be supported:
+- Construction
+  - `nil : () -> IntList`: construct an empty list
+  - `cons : (Int, IntList) -> IntList`: add a new item into the list
+- Deconstruction
+  - `head_opt : IntList -> Option[Int]`: get the first item
+  - `tail : IntList -> IntList`: get the rest items
+
+---
+
+# List Interfaces
+
+Test cases:
+
+```moonbit no-check
+let empty_list: IntList = nil()
+@assertion.assert_eq(head_opt(empty_list), None)?
+@assertion.assert_eq(tail(empty_list), empty_list)?
+
+let list1: IntList = cons(1, empty_list)
+@assertion.assert_eq(head_opt(list1), Some(1))?
+@assertion.assert_eq(tail(list1), empty_list)?
+
+let list2: IntList = cons(2, list1)
+@assertion.assert_eq(head_opt(list2), Some(2))?
+@assertion.assert_eq(tail(list2), list1)?
+```
+
+---
+
+# List in MoonBit
+
+- In the standard library of MoonBit, the list type is defined as:
+```moonbit
+enum List[T] {
+  Nil // an empty list
+  // or
+  Cons(T, List[T]) // an item of type T and a sublist whose items are also of type T
+}
+```
+
+- The definition of `List[T]` is inductive:
+  - Base case: `Nil`
+  - Inductive case: `Cons`
+
+![](../pics/list.drawio.png)
+
+---
+
+# Examples
+
+- The following are valid lists:
+    - `let int_list: List[Int] = Cons(1, Cons(2, Cons(3, Nil)))`
+    - `let string_list: List[String] = Cons("a", Cons("phrase", Nil))`
```

Comment:
Examples adjusted to accommodate non-Chinese readers.

Reply:
By the way, we have a syntax suger.
If a type `A` implements `A::from_array(array) -> A`, then we can write it with `A::[1, 2, 3]`. So you can write it with `List::[1, 2, 3]`, and `List::["a", "phrase"]`. Please introduce it somewhere.

Reply:
You don't need to introduce how it's done (I think we haven't introduced method yet), but you can say that we write `List::[1, 2, 3]` instead of `Cons(1, Cons(2, Cons(3, Nil)))`

Reply:
I added an introduction to this syntactic sugar in P16 and updated some examples in P17.